src/kudu/consensus/replicate_msg_wrapper.h

```diff
@@ -0,0 +1,193 @@
+#include "kudu/consensus/ref_counted_replicate.h"
+#include "kudu/gutil/strings/substitute.h"
+#include "kudu/util/compression/compression.pb.h"
+#include "kudu/util/compression/compression_codec.h"
+#include "kudu/util/crc.h"
+#include "kudu/util/faststring.h"
+
+namespace kudu {
+namespace consensus {
+
+/**
+ * Thin wrapper to handle compression/decompression of replicate msg
+ */
+class ReplicateMsgWrapper {
+public:
+  ReplicateMsgWrapper(const ReplicateRefPtr &msg,
+                      const CompressionCodec *codec_hint) {
+    orig_msg_ = msg;
+    const CompressionType msg_codec_type =
+        orig_msg_->get()->write_payload().compression_codec();
+    if (msg_codec_type == NO_COMPRESSION) {
+      msg_ = orig_msg_;
+      codec_ = codec_hint;
+    } else {
+      compressed_msg_ = orig_msg_;
+      CHECK_OK(GetCompressionCodec(msg_codec_type, &codec_));
+    }
+    should_compress_ =
+        orig_msg_->get()->op_type() == WRITE_OP_EXT && codec_ != nullptr;
+    DCHECK(msg_ || compressed_msg_);
```

Comment:
this behavior will be asymmetric between LEADERs and FOLLOWERs?
In the leader orig_msg_ will be uncompressed, and compressed_msg_ will be planted later?

Reply:
Yep

Reply:
we need to describe this behavior in detail.